apollo-router/src/plugins/file_uploads/mod.rs

```diff
@@ -0,0 +1,644 @@
+    }
+    Ok(req)
+}
+
+async fn subgraph_layer(mut req: subgraph::Request) -> subgraph::Request {
+    let supergraph_result = req
+        .context
+        .extensions()
+        .lock()
+        .get::<SupergraphLayerResult>()
+        .cloned();
+    if let Some(supergraph_result) = supergraph_result {
+        let SupergraphLayerResult {
+            multipart,
+            map_per_variable,
+            files_order,
+        } = supergraph_result;
+
+        let variables = &mut req.subgraph_request.body_mut().variables;
+        let mut map_field: MapField = IndexMap::new();
+        for (variable_name, variable_value) in variables.iter_mut() {
+            let variable_name = variable_name.as_str();
+            if let Some(variable_map) = map_per_variable.get(variable_name) {
+                for (file, paths) in variable_map.iter() {
+                    map_field.insert(
+                        file.clone(),
+                        paths
+                            .iter()
+                            .map(|path| {
+                                if path.is_empty() {
+                                    format!("variables.{}", variable_name)
+                                } else {
+                                    format!("variables.{}.{}", variable_name, path.join("."))
+                                }
+                            })
+                            .collect(),
+                    );
+                    for path in paths {
+                        if let Some(json_value) = try_path(variable_value, path) {
+                            json_value.take();
+                        }
+                    }
+                }
+            }
+        }
+        if !map_field.is_empty() {
+            map_field.sort_by_cached_key(|file, _| files_order.get_index_of(file));
+            req.subgraph_request
+                .extensions_mut()
+                .insert(SubgraphHttpRequestExtensions {
+                    multipart,
+                    map_field,
+                });
+        }
+    }
+    req
+}
+
+struct SubgraphHttpRequestExtensions {
+    multipart: MultipartRequest,
+    map_field: MapField,
+}
+
+const APOLLO_REQUIRE_PREFLIGHT: HeaderName = HeaderName::from_static("apollo-require-preflight");
+const TRUE: http::HeaderValue = HeaderValue::from_static("true");
+
+pub(crate) async fn http_request_wrapper(
+    mut req: http::Request<hyper::Body>,
+) -> http::Request<hyper::Body> {
+    let supergraph_result = req.extensions_mut().remove();
+    if let Some(supergraph_result) = supergraph_result {
+        let SubgraphHttpRequestExtensions {
+            multipart,
+            map_field,
+        } = supergraph_result;
+
+        let (mut request_parts, operations) = req.into_parts();
+        let form = MultipartFormData::new(operations, map_field, multipart);
+        request_parts
+            .headers
+            .insert(CONTENT_TYPE, form.content_type());
```

Comment:
Same question here for `insert` vs `append`

Reply:
We need to rewrite it from `application/json` to `multipart/form-data`.
I added comments in both places explaining this is intentional